File: modules/project/src/main/java/com/falsepattern/zigbrains/project/console/ZigSourceFileFilter.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.console;

import com.falsepattern.zigbrains.common.util.FileUtil;
import com.intellij.execution.filters.Filter;
import com.intellij.execution.filters.OpenFileHyperlinkInfo;
import com.intellij.openapi.project.Project;
import kotlin.Pair;
import lombok.RequiredArgsConstructor;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.file.Files;
import java.nio.file.InvalidPathException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.regex.Pattern;

@RequiredArgsConstructor
public class ZigSourceFileFilter implements Filter {
    private final Project project;
    private final Pattern LEN_REGEX = Pattern.compile(":(\\d+):(\\d+)");

    private Pair<Path, Integer> findLongestParsablePathFromOffset(String line, int end, String projectPath) {
        int longestStart = -1;
        Path longest = null;
        for (int i = end - 1; i >= 0; i--) {
            try {
                val pathStr = line.substring(i, end);
                var path = Path.of(pathStr);
                if (!Files.exists(path) || !Files.isRegularFile(path)) {
                    path = Path.of(projectPath, pathStr);
                    if (!Files.exists(path) || !Files.isRegularFile(path))
                        continue;
                }
                longest = path;
                longestStart = i;
            } catch (InvalidPathException ignored){}
        }
        return new Pair<>(longest, longestStart);
    }

    @Nullable
    @Override
    public Result applyFilter(@NotNull String line, int entireLength) {
        val lineStart = entireLength - line.length();
        val projectPath = project.getBasePath();
        val results = new ArrayList<ResultItem>();
        val matcher = LEN_REGEX.matcher(line);
        while (matcher.find()) {
            val end = matcher.start();
            val pair = findLongestParsablePathFromOffset(line, end, projectPath);
            val path = pair.getFirst();
            if (path == null)
                return null;
            val lineNumber = Math.max(Integer.parseInt(matcher.group(1)) - 1, 0);
            val lineOffset = Math.max(Integer.parseInt(matcher.group(2)) - 1, 0);
            val file = FileUtil.virtualFileFromURI(path.toUri());
            results.add(new ResultItem(lineStart + pair.getSecond(), lineStart + matcher.end(), new OpenFileHyperlinkInfo(project, file, lineNumber, lineOffset)));

        }
        return new Result(results);
    }
}

File: core/src/main/kotlin/com/falsepattern/zigbrains/project/console/ZigSourceFileFilter.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2025 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.console

import com.intellij.execution.filters.Filter
import com.intellij.execution.filters.Filter.ResultItem
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.guessProjectDir
import com.intellij.openapi.vfs.toNioPathOrNull
import java.io.File
import java.nio.file.InvalidPathException
import java.nio.file.Path
import kotlin.math.max


class ZigSourceFileFilter(private val project: Project): Filter {
    private val projectPath = runCatching { project.guessProjectDir()?.toNioPathOrNull()?.toFile() }.getOrNull()
    override fun applyFilter(line: String, entireLength: Int): Filter.Result? {
        if (line.isEmpty())
            return null
        val lineStart = entireLength - line.length
        val results = ArrayList<ResultItem>()
        var prevEnd = 0
        for (match in findLengthMarkers(line)) {
            val start = match.start
            val pair = findLongestParsablePathFromOffset(line, prevEnd, start) ?: return null
            prevEnd = match.end
            val path = pair.first
            val lineNumber = max(match.line - 1, 0)
            val lineOffset = max(match.offset - 1, 0)
            results.add(ResultItem(lineStart + pair.second, lineStart + match.end, LazyOpenFileHyperlinkInfo(project, path, lineNumber, lineOffset)))
        }
        if (results.isEmpty())
            return null
        return Filter.Result(results)
    }

    private fun findLongestParsablePathFromOffset(line: String, start: Int, end: Int): Pair<Path, Int>? {
        for (i in 0 ..< end) {
            try {
                val pathStr = line.substring(i, end)
                var file = File(pathStr)
                if (!file.isFile) {
                    if (projectPath == null) {
                        continue
                    }
                    file = projectPath.resolve(pathStr)
                    if (!file.isFile) {
                        continue
                    }
                }
                return Pair(file.toPath(), i)
            } catch (_: InvalidPathException) {
            }
        }
        return null
    }
}

private data class LengthMarker(val start: Int, val end: Int, val line: Int, val offset: Int)

private fun findLengthMarkers(line: String): Sequence<LengthMarker> = sequence {
    val chars = line.toCharArray()
    val len = chars.size
    var i = 0
    while (true) {
        while (chars[i] != ':') {
            i++
            if (i == len)
                return@sequence
        }
        val start = i
        i++
        if (i == len)
            return@sequence
        var line = 0
        while (chars[i] in '0'..'9') {
            line *= 10
            line += chars[i] - '0'
            i++
            if (i == len)
                return@sequence
        }
        if (chars[i] != ':')
            continue
        i++
        if (i == len)
            return@sequence
        var offset = 0
        while (chars[i] in '0'..'9') {
            offset *= 10
            offset += chars[i] - '0'
            i++
            if (i == len)
                return@sequence
        }
        yield(LengthMarker(start, i, line, offset))
    }
}
